From 4240bd1ee0f6fe498c1f4488f90ffb36076d57df Mon Sep 17 00:00:00 2001
From: Ramya Gnanasekar <quic_rgnanase@quicinc.com>
Date: Wed, 20 Apr 2022 11:23:28 +0530
Subject: [PATCH] ath12k: Skip SW decryption for groupcast mgmt frames

When protected bit is set, multicast/broadcast management frames will not
be decrypted by crypto HW and will be forwarded as encrypted payload
to the host.
In QCN9274, crypto hardware decrypts mutlicast/broadcast management
frames with protected bit set. Since host expects it to be encrypted
frame tires to decrypt again and fails. This causes mesh MPATH request which
is broadcast action frame get dropped.
Hence skipped the decryption routine for multicast/broadcast management
frames as well in case of PMF since it is already decrypted by crypto.

Signed-off-by: Ramya Gnanasekar <quic_rgnanase@quicinc.com>

diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index 7d3d92d..4f38c7e 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -6689,12 +6689,10 @@ static void ath12k_mgmt_rx_event(struct ath12k_base *ab, struct sk_buff *skb)
 	 */
 	status->flag |= RX_FLAG_SKIP_MONITOR;
 
-	/* In case of PMF, FW delivers decrypted frames with Protected Bit set.
-	 * Don't clear that. Also, FW delivers broadcast management frames
-	 * (ex: group privacy action frames in mesh) as encrypted payload.
+	/* In case of PMF, FW delivers decrypted frames with Protected Bit set
+	 * including group privacy action frames.
 	 */
-	if (ieee80211_has_protected(hdr->frame_control) &&
-	    !is_multicast_ether_addr(ieee80211_get_DA(hdr))) {
+	if (ieee80211_has_protected(hdr->frame_control)) {
 		status->flag |= RX_FLAG_DECRYPTED;
 
 		if (!ieee80211_is_robust_mgmt_frame(skb)) {
-- 
2.17.1

